libp2p/src/builder.rs

```diff
@@ -0,0 +1,1795 @@
+            },
+        }
+    }
+}
+
+impl<Provider, T: AuthenticatedMultiplexedTransport, R>
+    SwarmBuilder<Provider, WebsocketPhase<T, R>>
+{
+    fn without_websocket(self) -> SwarmBuilder<Provider, BandwidthLoggingPhase<T, R>> {
+        SwarmBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+            phase: BandwidthLoggingPhase {
+                relay_behaviour: self.phase.relay_behaviour,
+                transport: self.phase.transport,
+            },
+        }
+    }
+}
+
+// Shortcuts
+#[cfg(feature = "relay")]
+impl<Provider, T: AuthenticatedMultiplexedTransport>
+    SwarmBuilder<Provider, WebsocketPhase<T, libp2p_relay::client::Behaviour>>
+{
+    pub fn with_behaviour<B, R: TryIntoBehaviour<B>>(
+        self,
+        constructor: impl FnOnce(&libp2p_identity::Keypair, libp2p_relay::client::Behaviour) -> R,
+    ) -> Result<SwarmBuilder<Provider, SwarmPhase<T, B>>, R::Error> {
+        self.without_websocket()
+            .without_bandwidth_logging()
+            .with_behaviour(constructor)
+    }
+}
+
+impl<Provider, T: AuthenticatedMultiplexedTransport>
+    SwarmBuilder<Provider, WebsocketPhase<T, NoRelayBehaviour>>
+{
+    pub fn with_behaviour<B, R: TryIntoBehaviour<B>>(
+        self,
+        constructor: impl FnOnce(&libp2p_identity::Keypair) -> R,
+    ) -> Result<SwarmBuilder<Provider, SwarmPhase<T, B>>, R::Error> {
+        self.without_websocket()
+            .without_bandwidth_logging()
+            .with_behaviour(constructor)
+    }
+}
+
+#[cfg(all(not(target_arch = "wasm32"), feature = "websocket"))]
+pub struct WebsocketTlsPhase<T, R> {
+    transport: T,
+    relay_behaviour: R,
+}
+
+#[cfg(all(not(target_arch = "wasm32"), feature = "websocket"))]
+impl<Provider, T, R> SwarmBuilder<Provider, WebsocketTlsPhase<T, R>> {
+    #[cfg(feature = "tls")]
+    pub fn with_tls(
+        self,
+    ) -> Result<
+        SwarmBuilder<Provider, WebsocketNoisePhase<T, R, libp2p_tls::Config>>,
+        AuthenticationError,
+    > {
+        Ok(SwarmBuilder {
+            phase: WebsocketNoisePhase {
+                tls_config: libp2p_tls::Config::new(&self.keypair)
+                    .map_err(|e| AuthenticationError(e.into()))?,
+                relay_behaviour: self.phase.relay_behaviour,
+                transport: self.phase.transport,
+                phantom: PhantomData,
+            },
+            keypair: self.keypair,
+            phantom: PhantomData,
+        })
+    }
+
+    fn without_tls(self) -> SwarmBuilder<Provider, WebsocketNoisePhase<T, R, WithoutTls>> {
+        SwarmBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+            phase: WebsocketNoisePhase {
+                tls_config: WithoutTls {},
+                relay_behaviour: self.phase.relay_behaviour,
+                transport: self.phase.transport,
+                phantom: PhantomData,
+            },
+        }
+    }
+}
+
+// Shortcuts
```

Comment:
This file is very large. Can we split this into multiple modules? For example, I think all the shortcuts are not very important so they could all go in a module. All macros could go into a module too.